app/src/main/java/org/oppia/app/profile/PinPasswordActivityPresenter.kt

```diff
@@ -0,0 +1,141 @@
+import android.content.Intent
+import android.net.Uri
+import android.text.Editable
+import android.text.TextWatcher
+import androidx.appcompat.app.AlertDialog
+import androidx.appcompat.app.AppCompatActivity
+import androidx.databinding.DataBindingUtil
+import androidx.fragment.app.DialogFragment
+import androidx.lifecycle.Observer
+import org.oppia.app.R
+import org.oppia.app.databinding.PinPasswordActivityBinding
+import org.oppia.app.home.HomeActivity
+import org.oppia.app.model.ProfileId
+import org.oppia.app.viewmodel.ViewModelProvider
+import org.oppia.domain.profile.ProfileManagementController
+import javax.inject.Inject
+import android.view.animation.AnimationUtils
+import androidx.lifecycle.lifecycleScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+private const val TAG_ADMIN_SETTINGS_DIALOG = "ADMIN_SETTNIGS_DIALOG"
+private const val TAG_RESET_PIN_DIALOG = "RESET_PIN_DIALOG"
+
+class PinPasswordActivityPresenter @Inject constructor(
+  private val activity: AppCompatActivity,
+  private val profileManagementController: ProfileManagementController,
+  private val viewModelProvider: ViewModelProvider<PinPasswordViewModel>
+) {
+  private val pinViewModel by lazy {
+    getPinPasswordViewModel()
+  }
+  private var profileId = -1
+
+  @ExperimentalCoroutinesApi
+  fun handleOnCreate() {
+    val adminPin = activity.intent.getStringExtra(KEY_PIN_PASSWORD_ADMIN_PIN)
+    profileId = activity.intent.getIntExtra(KEY_PIN_PASSWORD_PROFILE_ID, -1)
+    val binding = DataBindingUtil.setContentView<PinPasswordActivityBinding>(activity, R.layout.pin_password_activity)
+    pinViewModel.setProfileId(profileId)
+    binding.apply {
+      lifecycleOwner = activity
+      viewModel = pinViewModel
+    }
+
+    binding.showPin.setOnClickListener {
+      pinViewModel.showPassword.set(!pinViewModel.showPassword.get()!!)
+    }
+
+    binding.inputPin.addTextChangedListener(object: TextWatcher {
+      var wrong = false
+      override fun onTextChanged(pin: CharSequence?, start: Int, before: Int, count: Int) {
+        pin?.let { inputtedPin ->
+          if (!wrong) {
+            pinViewModel.showError.set(false)
+          }
+          wrong = false
+          if (inputtedPin.length == pinViewModel.correctPin.length) {
+            if (inputtedPin.toString() == pinViewModel.correctPin) {
+              profileManagementController.loginToProfile(ProfileId.newBuilder().setInternalId(profileId).build())
+                .observe(activity, Observer {
+                if (it.isSuccess()) {
+                  activity.startActivity(Intent(activity, HomeActivity::class.java))
+                }
+              })
+            } else {
+              binding.inputPin.startAnimation(AnimationUtils.loadAnimation(activity, R.anim.shake))
+              activity.lifecycleScope.launch(Dispatchers.Main) {
+                delay(1000)
```

Comment:
We shouldn't be sleeping the main thread. This stops rendering which causes visual jankiness.

Reply:
Updated to Dispatchers.Default.

Reply:
Updated to use handler.postDelayed instead.